### PR TITLE
Incorporate SoA changes

### PR DESCRIPTION
**Pull request by ye-luo**
This PR is huge but it is better to merge it now than later.

SoA code paths are not activated by default.
No changes to the known issues and no new issues spotted so far.

To access SoA optimization, add -D ENABLE_SOA=1 in cmake.
Improvements:
Significant reduction of memory usage.
2X or more speed up on most CPU architectures.
Better mixed precision support.
Known issues:
Localized basis set support is not ready in SoA.

ToDo before next release:
1) Improve GNU compiler configurations and Cray handling.
2) Add the description of SoA in manual.
3) Enforce C++11 in CMake.

### CODE REVIEW

**Comment by ye-luo** [comment]
Okay to test

**Comment by prckent** [comment]
OK to test

**Comment by prckent** [comment]
How are local basis sets (Gaussians) handled with the SoA code? Abort, warnings, numerical problems?

**Comment by prckent** [comment]
Does this work with all the optimizers?

**Comment by prckent** [comment]
How badly broken do you estimate reptation to be? @rcclay needs it.

**Comment by prckent** [comment]
How does this interact with observables>

**Comment by ye-luo** [comment]
1) The SoA code aborts when seeing Gaussians.
2) Optimizers should work with SoA. OneShiftOnly is fully tested and others should be fine since changes are only in shared infrastructure.
3) I have fixed AoS RMC. SoA is broken and 1x16 shows terribly wrong results. But some 4x4 tests show pass. More observation needed.

**Comment by ye-luo** [comment]
Observables which are parts of the local energy are all checked.
MPC is checked. Observables catched by unit tests are all checked.
Basically, an observable relying on distance table should be checked.
The work process would be first creating a unit test and then fixing it.

**Comment by ye-luo** [comment]
@rcclay @prckent The RMC with SoA has been fixed.
I also noticed that the bccH RMC long test was not very robust.
Output from run without this PR.
----------------------------------------------------------
Tests for series 1
  Testing quantity: LocalEnergy
    reference mean value     :  -1.84150000
    reference error bar      :   0.00011000
    computed  mean value     :  -1.84199871
    computed  error bar      :   0.00010829
    pass tolerance           :   0.00033000  (  3.00000000 sigma)
    deviation from reference :  -0.00049871  ( -4.53368535 sigma)
    error bar of deviation   :   0.00015436
    significance probability :   0.99999420  (gaussian statistics)
    status of this test      :   fail

**Comment by rcclay** [comment]
Excellent.  Glad the fix came down to two lines.  I'll double check the RMC test.  When you say its not robust, how often does it fail?  Is it one of those things where you can restart it and it will pass, or is it failing somewhat frequently?

**Comment by ye-luo** [comment]
Real
https://cdash.qmcpack.org/CDash/testDetails.php?test=615077&build=8979
Complex
https://cdash.qmcpack.org/CDash/testDetails.php?test=615227&build=8980
Click  "Display graphs" Failing/Passing.
Seems about 20%~30%, though we don't have much statistics.

File: CMake/ClangCompilers.cmake [hunk 1]
```diff
@@ -5,7 +5,7 @@ IF ( CMAKE_CXX_COMPILER_VERSION VERSION_LESS 3.0 )
 ENDIF()
 
 # Enable OpenMP
-SET(ENABLE_OPENMP 0)
+SET(ENABLE_OPENMP 1)
```
Comment (markdewing) [inline CMake/ClangCompilers.cmake]: This might cause problems on Mac OS, since the clang there doesn't have OpenMP (the flag is silently ignored).   It could be fixed later.

Comment (ye-luo) [inline CMake/ClangCompilers.cmake]: Considering the increasing use of Clang on HPC, we might need to keep ENABLE_OPENMP as default and specifically protect the Mac if still necessary.

File: src/ParticleBase/ParticleBase.cpp [hunk 2]
```diff
@@ -323,79 +347,52 @@ ParticleBase<PL>::getTensorAttrib(const std::string& aname)
...
+  }
+
+  //curR.create(m);
   LocalNum += m;
   GlobalNum += m;
 }
 
 template<class PL>
-void ParticleBase<PL>::resize(unsigned m)
+void ParticleBase<PL>::resize(size_t m)
```
Comment (prckent) [inline src/ParticleBase/ParticleBase.cpp]: Much better

File: src/QMCApp/tests/test_wavefunction_pool.cpp [hunk 3]
```diff
@@ -47,12 +47,13 @@ const char *particles = \
       </parameter> \
         <parameter name='LR_dim_cutoff'>15 </parameter> \
 </simulationcell> \
-<particleset name=\"ion\" size=\"1\"> \
-  <group name=\"He\"> \
-    <parameter name=\"charge\">2</parameter> \
+<particleset name=\"ion\" size=\"2\"> \
```
Comment (prckent) [inline src/QMCApp/tests/test_wavefunction_pool.cpp]: What is the logic (concern) behind this change to the test?

File: src/QMCDrivers/VMC/VMCUpdateReny.cpp [hunk 4]
```diff
@@ -0,0 +1,569 @@
+//////////////////////////////////////////////////////////////////////////////////////
```
Comment (prckent) [inline src/QMCDrivers/VMC/VMCUpdateReny.cpp]: Is there a future plan to implement fully?

File: src/QMCHamiltonians/CoulombPotential.h [hunk 5]
```diff
@@ -117,35 +119,65 @@ struct CoulombPotential: public QMCHamiltonianBase
...
+            res+=q*Z[J[nn]]*d->rinv(nn);
+        }
       }
     }
     return res;
   }
 
 
-  inline T evaluateAB(const DistanceTableData* d, const ParticleScalar_t* restrict Za, const ParticleScalar_t* restrict Zb)
+  /** JNKIM: Need to check the precision */
```
Comment (prckent) [inline src/QMCHamiltonians/CoulombPotential.h]: Precision OK?

File: src/QMCHamiltonians/HamiltonianFactory.cpp [hunk 6]
```diff
@@ -808,13 +808,13 @@ bool HamiltonianFactory::put(xmlNodePtr cur)
...
   bool success=build(cur,false);
   
-  if(targetH->EnableVirtualMoves)
-  {
-    app_log() << "  Hamiltonian enables VirtualMoves" << std::endl;
-    targetPtcl->enableVirtualMoves();
-  }
-  else
-    app_log() << "  Hamiltonian disables VirtualMoves" << std::endl;
+  //if(targetH->EnableVirtualMoves)
```
Comment (prckent) [inline src/QMCHamiltonians/HamiltonianFactory.cpp]: Is this now reported elsewhere?

File: src/QMCHamiltonians/NonLocalECPComponent.cpp [hunk 7]
```diff
@@ -78,8 +78,12 @@ void NonLocalECPComponent::resize_warrays(int n,int m,int l)
   //for(int nl=1; nl<nlpp_m.size(); nl++) nlpp_m[nl]->setGridManager(false);
   if(lmax)
   {
-    Lfactor1.resize(lmax);
-    Lfactor2.resize(lmax);
+    if(lmax>7) 
+    {
+      APP_ABORT("Increase the maximum angular momentum in NonLocalECPComponet.h");
```
Comment (prckent) [inline src/QMCHamiltonians/NonLocalECPComponent.cpp]: Should say: increase the maximum angular momentum implemented

File: src/QMCWaveFunctions/Jastrow/TwoBodyJastrowOrbital.h [hunk 8]
```diff
@@ -744,6 +688,7 @@ class TwoBodyJastrowOrbital: public OrbitalBase
 
   RealType ChiesaKEcorrection()
   {
+#if QMC_BUILD_LEVEL<5
```
Comment (prckent) [inline src/QMCWaveFunctions/Jastrow/TwoBodyJastrowOrbital.h]: Why protect this part at high build levels?

File: src/formic/utils/lapack_interface.cpp [hunk 9]
```diff
@@ -76,6 +76,8 @@ double formic::ddot(int n, const double *x, int incx, const double *y, int incy)
 double formic::xdot(int n, const double *x, int incx, const double *y, int incy) {
   return ::FC_GLOBAL(ddot,DDOT)(&n, x, &incx, y, &incy);
 }
+
+/* commented to avoid compiler error.
```
Comment (prckent) [inline src/formic/utils/lapack_interface.cpp]: Which compiler? Good to leave a note.

File: src/miniapps/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,28 @@
+PROJECT(miniapps)
```
Comment (prckent) [inline src/miniapps/CMakeLists.txt]: Are we planning to maintain these miniapps?

File: src/miniapps/J2OrbitalSoA.Ye.h [hunk 11]
```diff
@@ -0,0 +1,505 @@
+//////////////////////////////////////////////////////////////////////////////////////
```
Comment (prckent) [inline src/miniapps/J2OrbitalSoA.Ye.h]: Why a .Ye version and a normal version?

File: src/simd/allocator.hpp [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+#ifndef QMCPLUSPLUS_ALLOCATOR_H
+#define QMCPLUSPLUS_ALLOCATOR_H
+
+#include <config.h>
+#include <vector>
+#include <cstdlib>
+
+#if (__cplusplus >= 201103L)
+#if defined(__INTEL_COMPILER)
+ #include <tbb/cache_aligned_allocator.h>
```
Comment (prckent) [inline src/simd/allocator.hpp]: TBB usage

File: src/simd/allocator.hpp [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+#else
+ #include "simd/Mallocator.hpp"
+#endif
+
+namespace qmcplusplus
+{
+  template<class T>
+#if defined(__INTEL_COMPILER)
+    using aligned_allocator=tbb::cache_aligned_allocator<T>;
```
Comment (prckent) [inline src/simd/allocator.hpp]: TBB usage

File: src/simd/allocator.hpp [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+#elif __bgq__
+    using aligned_allocator=std::allocator<T>;
```
Comment (prckent) [inline src/simd/allocator.hpp]: Why does BGQ need workaround here? Handle in Mallocator?

File: src/spline2/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,30 @@
+# add apps XYZ.cpp, e.g., qmc_particles.cpp
```
Comment (prckent) [inline src/spline2/CMakeLists.txt]: DIstinction between src/spline2 and src/spline?

Comment (prckent) [inline src/spline2/CMakeLists.txt]: Appears to be a redundant copy. Delete or fuse.

File: tests/system/CMakeLists.txt [hunk 16]
```diff
@@ -441,6 +451,17 @@ ENDIF()
                     0 # VMC
                     TRUE)
 
+IF(ENABLE_SOA)
```
Comment (prckent) [inline tests/system/CMakeLists.txt]: Why is the hybrid rep test dependent on SOA? Are there hybrid fixes not in the non-SOA code path?

File: CMake/FindMKL.cmake [hunk 17]
```diff
@@ -27,7 +27,7 @@ IF ( HAVE_MKL )
     SET( MKL_FOUND 1 )
     SET( MKL_FLAGS -mkl )
     SET( MKL_LIBRARIES )
-    SET( MKL_LINKER_FLAGS -mkl )
+    SET( MKL_LINKER_FLAGS -mkl -ltbbmalloc -ltbb)
```
Comment (ye-luo) [inline CMake/FindMKL.cmake]: This is active for Intel compiler only.
Does GNU+MKL hit here?

File: CMakeLists.txt [hunk 18]
```diff
@@ -182,10 +182,12 @@ CHECK_INCLUDE_FILE(unistd.h HAVE_UNISTD_H)
 # QMC_BUILD_STATIC build static/dynamic  libraries
 # BUILD_QMCTOOLS   build utility programs
 # BUILD_SANDBOX    build test programs
+# BUILD_MINIAPPS   build miniapps (Incomplete)
```
Comment (ye-luo) [inline CMakeLists.txt]: I can remove it.

File: src/OhmmsSoA/readme.md [hunk 19]
```diff
@@ -0,0 +1,39 @@
+ParticleAttribSoA
```
Comment (ye-luo) [inline src/OhmmsSoA/readme.md]: Better go for doxygen.

File: src/Particle/ParticleSet.cpp [hunk 20]
```diff
@@ -725,13 +774,17 @@ void ParticleSet::acceptMove(Index_t iat)
...
   {
     //Update position + distance-table
-    for (int i=0; i< DistTables.size(); i++)
-    {
+    for (int i=0,n=DistTables.size(); i< n; i++)
       DistTables[i]->update(iat);
-    }
+
+    if(RSoA.size() != getLocalNum())
+      std::cout << "Die here " << RSoA.size() << std::endl;
```
Comment (ye-luo) [inline src/Particle/ParticleSet.cpp]: It is debug lines.

File: src/QMCApp/tests/test_wavefunction_pool.cpp (hunk 3 above)
Comment (ye-luo) [inline src/QMCApp/tests/test_wavefunction_pool.cpp]: The original test pairs a He particle set with a Diamond h5 orbitals. Because of the hybrid rep code, I added a check ensuring atomic centres written in h5 must match xml input based on the tiling matrix.

File: src/QMCDrivers/VMC/VMCUpdateReny.cpp (hunk 4 above)
Comment (ye-luo) [inline src/QMCDrivers/VMC/VMCUpdateReny.cpp]: No idea, I guess this was fully implemented and may need very minor fix like RMC but need tests.

File: src/QMCHamiltonians/CoulombPotential.h (hunk 5 above)
Comment (ye-luo) [inline src/QMCHamiltonians/CoulombPotential.h]: This is open boundary condition. I only checked the O atom calculation in a big box. There is no fundamental change comparing to the old code. JK probably concerns about the accuracy of reciprocal in MP.

File: src/spline2/CMakeLists.txt (hunk 15 above)
Comment (ye-luo) [inline src/spline2/CMakeLists.txt]: Things in spline are helper functions for using einspline in EinsplineSet. Spline2 is the new 3D spline implementation to replace things in einspline folder. When we grave einspline, spline should also go away but not merged with spline2.

File: tests/system/CMakeLists.txt (hunk 16 above)
Comment (ye-luo) [inline tests/system/CMakeLists.txt]: Hybrid rep are build on top of new spline and new distance table implementation. No support to AoS.

File: src/simd/allocator.hpp (hunk 14 above)
Comment (ye-luo) [inline src/simd/allocator.hpp]: Mallocator is intended for machines having no aligned_allocator. BGQ std::allocator is aligned by default.

File: src/simd/allocator.hpp (hunk 12 above)
Comment (ye-luo) [inline src/simd/allocator.hpp]: Only for Intel compiler. To be replaced with future C++ standard solution.

File: src/simd/allocator.hpp (hunk 13 above)
Comment (ye-luo) [inline src/simd/allocator.hpp]: Only for Intel compiler. To be replaced with future C++ standard solution.

File: src/miniapps/J2OrbitalSoA.Ye.h (hunk 11 above)
Comment (ye-luo) [inline src/miniapps/J2OrbitalSoA.Ye.h]: My version has been in the Jastrow folder. This one is no more useful.

File: src/miniapps/CMakeLists.txt (hunk 10 above)
Comment (ye-luo) [inline src/miniapps/CMakeLists.txt]: They are harmless. Not build by default. I also use this place to host restart I/O miniapps.

File: src/QMCWaveFunctions/Jastrow/TwoBodyJastrowOrbital.h (hunk 8 above)
Comment (ye-luo) [inline src/QMCWaveFunctions/Jastrow/TwoBodyJastrowOrbital.h]: LEVEL=5 builds the executables in src/miniapps only. I think this was added for fair comparison between the old and new J2 implementation. In addition, based on Ray's claim, the Chiesa KE correction build on top of short range J2 is not theoretically correct and should be removed.

File: src/formic/utils/lapack_interface.cpp (hunk 9 above)
Comment (ye-luo) [inline src/formic/utils/lapack_interface.cpp]: My impression was Clang.

File: src/QMCHamiltonians/HamiltonianFactory.cpp (hunk 6 above)
Comment (ye-luo) [inline src/QMCHamiltonians/HamiltonianFactory.cpp]: I think this was no more used even in the AoS and the printout is no more relevant.

File: src/ParticleBase/ParticleBase.cpp (hunk 2 above)
Comment (ye-luo) [inline src/ParticleBase/ParticleBase.cpp]: ParticleBase will be deprecated. It is a layer not very necessary.

File: src/simd/allocator.hpp (hunk 13 above)
Comment (ye-luo) [inline src/simd/allocator.hpp]: This is basically a switching point for memory allocator and we can also add different allocator for different memory hierarchy. Mallocator is only a fall-back solution.

File: src/QMCDrivers/DMC/DMCOMP.cpp [hunk 21]
```diff
@@ -298,9 +283,16 @@ bool DMCOMP::run()
...
-        if ( step+1 == nSteps && nBlocksBetweenRecompute && (1+block)%nBlocksBetweenRecompute == 0 && QMCDriverMode[QMC_UPDATE_MODE] )
-          Movers[ip]->recomputePsi(wit,wit_end);
+#endif
+        wClones[ip]->resetCollectables();
+        const size_t nw=W.getActiveWalkers();
+#pragma omp for nowait
+        for(size_t iw=0;iw<nw; ++iw)
+        {
+          Walker_t& thisWalker(*W.getWalker(iw));
+          Movers[ip]->advanceWalker(thisWalker,recompute);
```
Comment (rcclay) [inline src/QMCDrivers/DMC/DMCOMP.cpp]: Why did we change from iterating over walkers within "advanceWalker" to loading single walkers in at a time?

File: src/Particle/MCWalkerConfiguration.h [hunk 22]
```diff
@@ -211,6 +211,11 @@ class MCWalkerConfiguration: public ParticleSet
     return R.size();
   }
 
+  inline Walker_t* getWalker(size_t iw)
```
Comment (ye-luo) [inline src/Particle/MCWalkerConfiguration.h]: You are right. Remove getWalker and use [].

File: src/Particle/ParticleSet.h [hunk 23]
```diff
@@ -285,26 +294,23 @@ class ParticleSet
   }
 
   /** update the internal data
-   *@param iflag index for the update mode
+   *@param skip SK update if skipSK is true
    */
-  void update(int iflag=0);
+  void update(bool skipSK=false);
```
Comment (ye-luo) [inline src/Particle/ParticleSet.h]: update() is not supposed to be called often.
donePbyP() computes SK before any estimators.

File: src/QMCDrivers/DMC/DMCOMP.cpp (hunk 21 above)
Comment (ye-luo) [inline src/QMCDrivers/DMC/DMCOMP.cpp]: This is to keep DMC/VMCUpdatePbyP/All. The manipulation of walkers is all handled in the direct VMC/DMC driver.

File: CMake/FindMKL.cmake (hunk 17 above)
Comment (ye-luo) [inline CMake/FindMKL.cmake]: Removed TBB

File: src/Particle/ParticleSet.cpp (hunk 20 above)
Comment (ye-luo) [inline src/Particle/ParticleSet.cpp]: Removed.

File: src/QMCHamiltonians/NonLocalECPComponent.cpp (hunk 7 above)
Comment (ye-luo) [inline src/QMCHamiltonians/NonLocalECPComponent.cpp]: Updated

File: CMakeLists.txt (hunk 18 above)
Comment (prckent) [inline CMakeLists.txt]: OK

File: src/OhmmsSoA/readme.md (hunk 19 above)
Comment (prckent) [inline src/OhmmsSoA/readme.md]: For now, and the likely future, we are putting developer notes in the main manual to make them more accessible and less likely to be forgotten about.

File: src/ParticleBase/ParticleBase.cpp (hunk 2 above)
Comment (prckent) [inline src/ParticleBase/ParticleBase.cpp]: OK

File: src/QMCApp/tests/test_wavefunction_pool.cpp (hunk 3 above)
Comment (prckent) [inline src/QMCApp/tests/test_wavefunction_pool.cpp]: Good point.

File: src/QMCDrivers/VMC/VMCUpdateReny.cpp (hunk 4 above)
Comment (prckent) [inline src/QMCDrivers/VMC/VMCUpdateReny.cpp]: OK. Unless someone claims ownership and produces tests and inputs, this code is likely to be removed shortly.

File: src/QMCHamiltonians/CoulombPotential.h (hunk 5 above)
Comment (prckent) [inline src/QMCHamiltonians/CoulombPotential.h]: We will leave this comment, but in future (and in general) developers should either write a longer comment about the concern that will be meaningful to others or make the checks and remove the comment...   (not targeting JK here)

File: src/QMCHamiltonians/HamiltonianFactory.cpp (hunk 6 above)
Comment (prckent) [inline src/QMCHamiltonians/HamiltonianFactory.cpp]: OK

File: src/QMCWaveFunctions/Jastrow/TwoBodyJastrowOrbital.h (hunk 8 above)
Comment (prckent) [inline src/QMCWaveFunctions/Jastrow/TwoBodyJastrowOrbital.h]: OK.

File: src/formic/utils/lapack_interface.cpp (hunk 9 above)
Comment (prckent) [inline src/formic/utils/lapack_interface.cpp]: Note to future devs: *please* write sufficiently meaningful comments. Which compiler? Which version? Which options?

File: src/miniapps/CMakeLists.txt (hunk 10 above)
Comment (prckent) [inline src/miniapps/CMakeLists.txt]: OK if needed for I/O work.

File: src/spline2/CMakeLists.txt (hunk 15 above)
Comment (prckent) [inline src/spline2/CMakeLists.txt]: We need to tiny the many splines. This is a fine example of how to reduce accessibility of a code.

File: tests/system/CMakeLists.txt (hunk 16 above)
Comment (prckent) [inline tests/system/CMakeLists.txt]: Clear now.

File: src/Particle/ParticleSet.h (hunk 23 above)
Comment (prckent) [inline src/Particle/ParticleSet.h]: OK

File: src/QMCDrivers/DMC/DMCOMP.cpp (hunk 21 above)
Comment (prckent) [inline src/QMCDrivers/DMC/DMCOMP.cpp]: OK.

File: src/simd/allocator.hpp (hunk 14 above)
Comment (prckent) [inline src/simd/allocator.hpp]: OK

**Comment by prckent** [comment]
I am going to approve this PR and will merge tomorrow if no one comments to delay further.

Thanks to everyone at Intel, Sandia, Argonne who contributed the impressive performance improvements and new implementations.

To manage the complexity -- SoA is essentially an internal fork in the code, just like the CUDA implementation -- we'll have to update the remaining important features to SoA after first defining tests for them. Unclaimed parts of the code AoS code will then become part of git history. This is a very likely target for release v4.0.0

**Comment by prckent** [comment]
Merging. Please note that this version uses a newer spline implementation even with ENABLE_SOA=0 (default).
